core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala

```diff
@@ -119,18 +119,43 @@ private[ui] object RDDOperationGraph extends Logging {
       { if (stage.attemptId == 0) "" else s" (attempt ${stage.attemptId})" }
     val rootCluster = new RDDOperationCluster(stageClusterId, stageClusterName)
 
+    var rootNodeCount = 0
+    val addRDDIds = new mutable.HashSet[Int]()
+    val dropRDDIds = new mutable.HashSet[Int]()
+
+    def isAllowed(rdd: RDDInfo): Boolean = {
+      val parentIds = rdd.parentIds
+      if (parentIds.size == 0) {
+        rootNodeCount < retainedNodes
+      } else {
+        parentIds.exists(id => addRDDIds.contains(id) || !dropRDDIds.contains(id))
+      }
+    }
+
     // Find nodes, edges, and operation scopes that belong to this stage
-    stage.rddInfos.foreach { rdd =>
-      edges ++= rdd.parentIds.map { parentId => RDDOperationEdge(parentId, rdd.id) }
+    stage.rddInfos.sortBy(_.id).foreach { rdd =>
+      val keepNode = isAllowed(rdd)
+      if (keepNode) {
+        addRDDIds.add(rdd.id)
```

Comment:
Hm, maybe you can't because it depends on the set's current value. Anyway I wonder if you can just inline isAllowed to make this simpler.